python/ELASTICACHE_REDIS_CLUSTER_AUTO_BACKUP_CHECK/ELASTICACHE_REDIS_CLUSTER_AUTO_BACKUP_CHECK.py

```diff
@@ -0,0 +1,444 @@
+     Given: At least 1 Amazon Elasticache Redis cluster is present
+       And: The SnapshotRetentionLimit is set to 0
+      Then: Return NON_COMPLIANT with Annotation "Automatic backup not enabled for Amazon ElastiCache cluster {Cluster_ID}"
+  Scenario: 4
+     Given: At least 1 Amazon Elasticache Redis cluster is present
+       And: The SnapshotRetentionLimit is less than snapshotRetentionPeriod
+      Then: Return NON_COMPLAINT with Annotation "Automatic backup retention period for Amazon ElastiCache cluster {Cluster_ID} is less then {snapshotRetentionPeriod} day(s)."
+  Scenario: 5
+     Given: At least 1 Amazon Elasticache Redis cluster is present
+       And: The SnapshotRetentionLimit is greater than or equal to snapshotRetentionPeriod
+      Then: Return COMPLAINT
+'''
+import json
+import sys
+import datetime
+import boto3
+import botocore
+
+try:
+    import liblogging
+except ImportError:
+    pass
+
+##############
+# Parameters #
+##############
+
+# Define the default resource to report to Config Rules
+DEFAULT_RESOURCE_TYPE = 'AWS::::Account'
+
+# Set to True to get the lambda to assume the Role attached on the Config Service (useful for cross-account).
+ASSUME_ROLE_MODE = False
+
+# Other parameters (no change needed)
+CONFIG_ROLE_TIMEOUT_SECONDS = 900
+
+#############
+# Main Code #
+#############
+
+
+def get_replication_groups(ec_client):
+    replication_groups = []
+    marker = None
+    replication_groups_result = {}
+    while True:
+        if not marker:
+            replication_groups_result = ec_client.describe_replication_groups(MaxRecords=100)
+        else:
+            replication_groups_result = ec_client.describe_replication_groups(Marker=marker, MaxRecords=100)
+        replication_groups.extend(replication_groups_result['ReplicationGroups'])
+        if 'Marker' in replication_groups_result:
+            marker = replication_groups_result['Marker']
+        else:
+            return replication_groups
+
+
+def get_cache_clusters(ec_client):
+    cache_clusters = []
+    marker = None
+    cache_clusters_result = {}
+    while True:
+        if not marker:
+            cache_clusters_result = ec_client.describe_cache_clusters(MaxRecords=100, ShowCacheNodeInfo=False, ShowCacheClustersNotInReplicationGroups=True)
+        else:
+            cache_clusters_result = ec_client.describe_cache_clusters(Marker=marker, MaxRecords=100, ShowCacheNodeInfo=False, ShowCacheClustersNotInReplicationGroups=True)
+        cache_clusters.extend(cache_clusters_result['CacheClusters'])
+        if 'Marker' in cache_clusters_result:
+            marker = cache_clusters_result['Marker']
+        else:
+            return cache_clusters
+
+
+def generate_evaluations(cache_clusters, replication_groups, snapshot_retention_period, event):
+    evaluations = []
+    if cache_clusters:
+        for cluster in cache_clusters:
+            if cluster['Engine'] == 'redis':
+                if cluster['SnapshotRetentionLimit'] == 0:

```

Comment:
Too many if/else. Please use guardclauses. https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html

Reply:
I moved the engine check to get_cache_clusters function so we do not need to perform this check